src/transform/src/analysis/equivalences.rs

```diff
@@ -0,0 +1,560 @@
+                    let mut cloned = self.classes[class_index][index].clone();
+                    // Use `reduce_child` rather than `reduce_expr` to avoid entire expression replacement.
+                    let reduced = self.reduce_child(&mut cloned);
+                    if reduced {
+                        self.classes[class_index][index] = cloned;
+                        complete = false;
+                    }
+                }
+            }
+
+            // 2. Unify classes.
+            //    If the same expression is in two classes, we can unify the classes.
+            //    This element may not be the representative.
+            //    TODO: If all lists are sorted, this could be a linear merge among all.
+            //          They stop being sorted as soon as we make any modification, though.
+            //          But, it would be a fast rejection when faced with lots of data.
+            for index1 in 0..self.classes.len() {
+                for index2 in 0..index1 {
+                    if self.classes[index1]
+                        .iter()
+                        .any(|x| self.classes[index2].iter().any(|y| x == y))
+                    {
+                        let prior = std::mem::take(&mut self.classes[index2]);
+                        self.classes[index1].extend(prior);
+                        complete = false;
+                    }
+                }
+            }
+
+            // 3. Identify idioms
+            //    E.g. If Eq(x, y) must be true, we can introduce classes `[x, y]` and `[true, Not(IsNull(x), Not(IsNull(y))]`.
+            let mut to_add = Vec::new();
+            for class in self.classes.iter_mut() {
+                if class.iter().any(|c| c.is_literal_true()) {
+                    for expr in class.iter() {
+                        if let MirScalarExpr::CallBinary {
+                            func: mz_expr::BinaryFunc::Eq,
+                            expr1,
+                            expr2,
+                        } = expr
+                        {
+                            to_add.push(vec![*expr1.clone(), *expr2.clone()]);
+                            to_add.push(vec![
+                                MirScalarExpr::literal_true(),
+                                expr1.clone().call_is_null().not(),
+                                expr2.clone().call_is_null().not(),
+                            ]);
+                        }
+                    }
+                    class.retain(|expr| {
+                        if let MirScalarExpr::CallBinary {
+                            func: mz_expr::BinaryFunc::Eq,
+                            expr1: _,
+                            expr2: _,
+                        } = expr
+                        {
+                            false
+                        } else {
+                            true
+                        }
+                    });
+                }
+            }
+            self.classes.extend(to_add);
```

Comment:
Missing `complete = false;` when `to_add` is not empty. E.g., `Not(IsNull(x)` might be further simplified by step 0.